### PR TITLE
[RLlib] DreamerV3: RLModule class, tf.keras model components, dreamer-model, world-model, actor, critic

### PR DESCRIPTION
**Pull request by sven1977**
DreamerV3: RLModule class, tf.keras model components, dreamer-model, world-model, actor, critic

Note: This PR is non-functional in itself. It merely introduces the individual models and model components (tf keras) needed to run the DreamerV3 algorithm, of which integration into RLlib is in progress (separate PRs will be released and merged in the next days).

cc: @danijar :) 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/dreamerv3/tf/models/components/representation_layer.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+
+if __name__ == "__main__":
+    layer = RepresentationLayer(num_categoricals=32, num_classes_per_categorical=32)
+    # encoder output
+    x = np.random.random(size=(1, 128))
+    # GRU output
+    h = np.random.random(size=(1, 512))
+    out = layer(tf.concat([x, h], axis=-1))
+    print(out.shape)
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/tf/models/components/representation_layer.py:139]: remove this or create a unittest?

File: rllib/algorithms/dreamerv3/tf/models/disagree_networks.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            )
+
+        # Optimizer.
+        self.optimizer = tf.keras.optimizers.Adam(learning_rate=1e-4, epsilon=1e-5)
+
+    # @tf.functionction
+    def call(self, inputs, z, a, training=None):
+        return self.forward_train(a=a, h=inputs, z=z)
+
+    # @tf.functionction
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/tf/models/disagree_networks.py:55]: remove?

File: rllib/algorithms/dreamerv3/tf/models/disagree_networks.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                )
+            )
+            self.representation_layers.append(
+                RepresentationLayer(
+                    model_dimension=self.model_dimension, name="disagree"
+                )
+            )
+
+        # Optimizer.
+        self.optimizer = tf.keras.optimizers.Adam(learning_rate=1e-4, epsilon=1e-5)
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/tf/models/disagree_networks.py:49]: should we remove optimizer from this model ?

File: rllib/core/rl_module/tf/tf_rl_module.py [hunk 4]
```diff
@@ -8,8 +8,8 @@
 _, tf, _ = try_import_tf()
 
 
-class TfRLModule(RLModule, tf.keras.Model):
```
Comment (kouroshHakha) [inline rllib/core/rl_module/tf/tf_rl_module.py:11]: What broke with the ordering?

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+from ray.rllib.core.models.specs.specs_dict import SpecDict
+from ray.rllib.core.rl_module.rl_module import RLModule
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import ExperimentalAPI, override
+from ray.rllib.utils.numpy import one_hot
+
+
+@ExperimentalAPI
+class DreamerV3RLModule(RLModule, abc.ABC):
+    def setup(self):
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:24]: override decorator?

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
+"""
+This file holds framework-agnostic components for DreamerV3's RLModule.
+"""
+
+import abc
+
+import numpy as np
+
+from ray.rllib.algorithms.dreamerv3.utils import do_symlog_obs
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:9]: do_symlog_obs is missing, right? the utils it not part of this PR.

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            reps=(B, T, 1),
+        )
+        self.dreamer_model(
+            inputs=test_obs,
+            actions=test_actions.astype(np.float32),
+            is_first=np.ones((B, T), np.float32),
+            start_is_terminated_BxT=np.zeros((B * T,), np.float32),
+            horizon_H=horizon_H,
+            gamma=gamma,
+        )
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:94]: wait, why is this needed?

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        # This should work now.
+        self.dreamer_model.summary(expand_nested=True)
+
+        # Initialize the critic EMA net:
+        self.critic.init_ema()
+
+    # @override(RLModule)
+    # def get_initial_state(self) -> NestedDict:
+    #    return self.dreamer_model.get_initial_state()
```
Comment (kouroshHakha) [inline rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:103]: delete?

**Review by kouroshHakha — APPROVED** [review]

File: rllib/core/rl_module/tf/tf_rl_module.py (hunk 4 above)
Comment (sven1977) [reply rllib/core/rl_module/tf/tf_rl_module.py:11]: I'm actually not 100% sure yet. Here are my observations thus far:
- TorchRLModule also has nn.Module first, then RLModule (why?, not sure)
- If we put tf.keras.Model last, then we must specify a proper init c'tor for TfRLModule, b/c the c'tor of keras.Model must be called first (otherwise, keras complains).
- However, if we do write our own c'tor ...
```
def __init__(self, config):
    tf.keras.Model.__init__(self)
    RLModule.__init__(self, config)
```
... then - for some reason - the `setup()` method is called twice and we create two(!) actual modules under the hood.
- My thought then was: put tf.keras.Model first, scrap the c'tor altogether (then python would call the super c'tors in the correct order w/o the duplicate `setup()` call), and we should be good.
- It's this weird multiple-inheritance thing in combination with not using `super().__init__` that screws up things here, I think.

I'll try to poke around and find a clean solution for this.

File: rllib/algorithms/dreamerv3/tf/models/disagree_networks.py (hunk 3 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/models/disagree_networks.py:49]: Great catch. Actually, let me remove the entire disagree stuff. I haven't explored this well yet anyways. We can re-add this functionality later from my DreamerV3 repo, if needed.

File: rllib/algorithms/dreamerv3/tf/models/disagree_networks.py (hunk 2 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/models/disagree_networks.py:55]: Ah, sorry, yes, these should all be removed, I guess. eager tracing is now entirely done via RLlib/RLModules.

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py (hunk 7 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:94]: To define all variables in the module. If I don't make this call, module.trainable_variables will be empty and I won't be able to print out a summary of the model, which is super useful to see what we have actually built.

File: rllib/algorithms/dreamerv3/tf/models/disagree_networks.py (hunk 2 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/models/disagree_networks.py:55]: done

File: rllib/algorithms/dreamerv3/tf/models/components/dynamics_predictor.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        return self.representation_layer(out, return_z_probs=return_z_probs)
+
+
+if __name__ == "__main__":
+    # DreamerV2/3 Atari input space: B x 32 (num_categoricals) x 32 (num_classes)
+    h_dim = 8
+    inputs = np.random.random(size=(1, 8))
+    model = DynamicsPredictor()
+    out = model(inputs)
+    print(out.shape)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tf/models/components/dynamics_predictor.py:82]: done

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py (hunk 5 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:24]: great catch! done

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py (hunk 8 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:103]: Moved to framework-agnostic super class and deleted the commented-out code

File: rllib/core/rl_module/tf/tf_rl_module.py (hunk 4 above)
Comment (kouroshHakha) [reply rllib/core/rl_module/tf/tf_rl_module.py:11]: Can we then add comments on the specifics of this choice that we made? This will become a wuestion later again.

Comment (sven1977) [reply rllib/core/rl_module/tf/tf_rl_module.py:11]: I'll try to clean this up better (make torch and tf the same) and add an explanation. I'm just worried that we already chose a quite complex inheritance structure and this is what's causing the `setup()` to get called twice. Will figure it out ...

Comment (kouroshHakha) [reply rllib/core/rl_module/tf/tf_rl_module.py:11]: That's probably the case. But as long as it's documented it should be fine 😀

File: rllib/algorithms/dreamerv3/dreamerv3_rl_module.py (hunk 7 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3_rl_module.py:96]: An alternative to making this call would be to provide input shape information to each of the components, which, to be honest, is actually more work than performing this single call with proper test data. This seems to be a tf only problem as for torch, you would have to provide input shapes anyways when you build your nn.Modules.
However, if you don't make the call, the tf.Model is not considered "built" and we won't be able to do `summary()`, even if the variables are known. I agree, we should not print out the summary in general. Maybe just on the driver and only in debug logging mode.
